buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java

```diff
@@ -1815,7 +1815,11 @@ ByteBuf duplicate() {
         }
 
         void freeIfNecessary() {
-            buf.release(); // We should not get a NPE here. If so, it must be a bug.
+            // Release the slice if present since it may have a different
+            // refcount to the unwrapped buf if it is a PooledSlicedByteBuf
+            ByteBuf s = slice;
+            (s != null ? s : buf).release();
+            // null out in both cases since it could be racy
```

Comment:
I think this would be easier to read like this:

```
ByteBuf buffer = s == null ? buf : s;
buffer.release();
```

Reply:
@njhill I am confused... why did you need to to have `s` and `toRelease` here ? Which not just do what I suggested ? 

Which would be:

```
ByteBuf buffer = slice == null ? buf : slice;
buffer.release();
```

Reply:
@normanmaurer oh, sorry... I misinterpreted your original suggestion since it used `s`. Maybe I am being overly cautious, but just based on the fact that `slice` gets nulled after the component is released I thought it would be safer to do the check on a local var (I ack that it probably is only possible to hit an NPE problem if there is another bug or incorrect use though).

Happy to change it to use `slice` directly if you think that's fine...

Reply:
yeah I think I would go for the easier code or would at least not write it so compact to make it easier to understand

Reply:
done - changed to use if/else, hopefully this is the most readable of all :)